pytest_variables/plugin.py

```diff
@@ -49,23 +50,49 @@ def pytest_addoption(parser):
         help='path to variables file.')
 
 
+def _merge(a, b, path=None):
+    """ merges b and a configurations.
+        Based on http://bit.ly/2uFUHgb
+     """
+    if path is None:
+        path = []
+
+    for key in b:
+        if key in a:
+            if isinstance(a[key], dict) and isinstance(b[key], dict):
+                _merge(a[key], b[key], path + [str(key)])
+            elif a[key] == b[key]:
+                pass  # same leaf value
+            else:
+                # b wins
+                a[key] = b[key]
+        else:
+            a[key] = b[key]
+    return a
+
+
 def pytest_configure(config):
     config._variables = {}
-    for path in config.getoption('variables'):
+    paths = config.getoption('variables')
+    for path in paths:
         ext = os.path.splitext(path)[1][1:].lower() or 'json'
         try:
             variables = import_parser(path, *parser_table[ext])
-            config._variables.update(variables)
         except KeyError:
             print("Could not find a parser for the file extension '{0}'. "
                   'Supported extensions are: {1}'.format(
                       ext, ', '.join(sorted(parser_table.keys()))))
-            config._variables.update(
-                import_parser(path, *parser_table['json']))
+            variables = import_parser(path, *parser_table['json'])
         except ValueError as e:
             raise errors.ValueError('Unable to parse {0}: {1}'.format(
                 path, e))
 
+        if not isinstance(variables, dict):
+            raise errors.ValueError('Unable to parse {0}'.format(
+                path))
+
+        reduce(_merge, [config._variables, variables])
```

Comment:
we can avoid the loop ant let the reduce update config._variables with just one call ``reduce(_merge, [config._variables, variables1, ...])``. I decided to use a loop in order to keep things backwards compatible with existing tests